### PR TITLE
[MD] Register data source plugin interface to route handler context

### LINKED ISSUE #1979: [MD] Register getClient interface from `data_source` plugin to request context as plugin context

### PR DESCRIPTION
**Pull request by zhongnansu**
Signed-off-by: Zhongnan Su <szhongna@amazon.com>

### Description
- Create data source plugin context class that contains the interface to retrieve a client by datasource id.
- Register data source plugin context to route handler context. 
- The caller plugin, for example, search plugin will use the interface as below to retrieve a data source client. It will be further implementer in other tasks. #1988
```
const dataSourceClient = await context.dataSource.opensearch.getClient(id)
``` 
 - Init `DataSourceClient` class, with a dummy implementation of functions required by interfaces , it will be further implemented in other PRs
 - Init `DataSourceService` class, with interfaces defined
### Issues Resolved
#1979 
 #1980
#1984 
#1982 
#1983
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
    - [ ] `yarn test:jest`
    - [ ] `yarn test:jest_integration`
    - [ ] `yarn test:ftr`
- [ ] New functionality has been documented.
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (noCharger) [inline src/plugins/data_source/server/client/data_source_client.ts:40]: nit: remove comment

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (noCharger) [inline src/plugins/data_source/server/client/data_source_client.ts:10]: can you help me understand the purpose of 'user' in the 'data source' saved object? I didn't see it in the current implementation.

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (noCharger) [inline src/plugins/data_source/server/data_source_route_handler_context.ts:25]: Just curious, what could go wrong when retrieving an opensearch client?

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (kristenTian) [inline src/plugins/data_source/server/client/data_source_client.ts:31]: why need a separate interface to include close method but not in IDataSourceClient?

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (kristenTian) [inline src/plugins/data_source/server/data_source_route_handler_context.ts:12]: add dataSourceClient to class field as well to be consistent?

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (kristenTian) [inline src/plugins/data_source/server/data_source_route_handler_context.ts:30]: what is the need of this wrapper? Why not export `OpenSearchDataSourceRouteHandlerContext` directly?

File: src/plugins/data_source/server/plugin.ts
Comment (kristenTian) [inline src/plugins/data_source/server/plugin.ts:30]: why add async here? seems we are just returning an async function.

File: src/plugins/data_source/server/plugin.ts
Comment (kristenTian) [inline src/plugins/data_source/server/plugin.ts:60]: dummy question, how is DataSourceRouteHandlerContext related to IContextProvider? also is the name mis-matching: ...contextProvider vs ...context

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/client/data_source_client.ts:10]: I mean the user/credential associated with data source

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/client/data_source_client.ts:31]: The `close()` function is a critical one that closes all the connections in pool, that we don't want to expose this function to data source client consumers, who only needs to retrieve the client

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_route_handler_context.ts:12]: I'll remove the logger, to use parameter property here for consistency. 
https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/client/data_source_client.ts:31]: close() should only be called when OSD or plugin is shut down

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts:30]: good catch, will fix that

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts:60]: We provide our plugin context as a request handler context. `DataSourceRouteHandlerContext` contains the property we want to register. The register function requires to pass in a provider, which you can see is a type of `IContextProvider`

```
registerRouteHandlerContext: <T extends keyof RequestHandlerContext>(
    contextName: T,
    provider: RequestHandlerContextProvider<T>v
  ) => RequestHandlerContextContainer;

export type RequestHandlerContextProvider<
  TContextName extends keyof RequestHandlerContext
> = IContextProvider<RequestHandler<any, any, any>, TContextName>;
```

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_route_handler_context.ts:30]: Just making this flexible, maybe in the future we add other types of datas source, and this also following the `core_route_handler_context.ts` manner.

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_route_handler_context.ts:25]: Connection error, time out, authentication failed, things like that

**Review by kristenTian — APPROVED** [review]

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:27]: by having one single `DataSourceClient` object, and you may update the saved object client in that object for each request, will it causing racing condition and mess up the DataSourceClient object internal state?

**Review by noCharger — APPROVED** [review]
LGTM, would like to see comment from @zengyan-amazon  resolved before merging

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts]: I am actually not sure, cuz I thought node.js is single threaded, and safe. But anyways, I added another layer of abstraction "dataSourceService", which has function to create `dataSourceClient` with `scopedSavedObjectClient` on demand,  could you take another look? I feel that makes more sense

fyi @noCharger

File: src/plugins/credential_management/public/components/breadcrumbs.ts
Comment (zengyan-amazon) [inline src/plugins/credential_management/public/components/breadcrumbs.ts:32]: why updating credential management UI? let's do a separate PR for management UI changes

File: src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts
Comment (zengyan-amazon) [inline src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts:96]: seems some other changes is added to this PR, can you clean it up?

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (zengyan-amazon) [reply src/plugins/data_source/server/client/data_source_client.ts:32]: where is this `close()` function being used?

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/data_source_route_handler_context.ts:34]: `logger.get('opensearch')` why using `opensearch` logger? will it has conflict with OpenSearchService?

File: src/plugins/data_source/server/data_source_service.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/data_source_service.ts:18]: as discussed, let's use datasource plugin enablement as the global config

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:60]: you can use `context.savedObjects.client`

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:56]: move this function to context related file

File: src/plugins/data_source/server/data_source_service.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/data_source_service.ts:34]: let's include the implementation

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/data_source_route_handler_context.ts:29]: need to handle legacy client as well

**Review by zengyan-amazon — CHANGES_REQUESTED** [review]

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [reply src/plugins/data_source/server/plugin.ts]: NodeJs is single threaded, but the request processing is async, so you still have the issue when your class internal state is messed up

File: src/plugins/credential_management/public/components/breadcrumbs.ts
Comment (zhongnansu) [reply src/plugins/credential_management/public/components/breadcrumbs.ts:32]: That's not my change. Upstream had a force push that leads to some diffs. I'll fix it

File: src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts
Comment (zhongnansu) [reply src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts:96]: Upstream had a force push that leads to some diffs. I'll fix it

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_route_handler_context.ts:34]: I want to mark the logger type for datasource as an pensearch data source. It's the sub-logger of plugin logger. So I don't think it will mess up. The log format from here will look like
[DataSource][Plugin][opensearch]:xxxxx

File: src/plugins/data_source/server/data_source_service.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_service.ts:18]: yes, this will be aligned with global config.

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts:60]: yeah you are right, let me fix that

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts:56]: will do

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/client/data_source_client.ts:32]: I forget to remove this. So after adding `DataSourceService` , we'll handle closing connections at the service level. Thanks for pointing it out

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_route_handler_context.ts:29]: What's the usage of legacy openesarch client currently? I didn't think too much there as I thought the critical data source consumer,  "search" plugin,  is not relying on legacy clients. Also as I see [here](https://www.elastic.co/guide/en/kibana/7.17/migrating-legacy-plugins.html), the support of legacy is removed after 7.10 for plugins?

> In Kibana 7.10, support for legacy-style Kibana plugins was completely removed. Moving forward, all plugins must be built on the new Kibana Platform Plugin API. This guide is intended to assist plugin authors in migrating their legacy plugin to the Kibana Platform Plugin API.

But I did find some usage through code referencing, still need to dive deep on those. I can create an issue for tracking.
image

File: src/plugins/data_source/server/data_source_service.ts
Comment (noCharger) [reply src/plugins/data_source/server/data_source_service.ts:19]: we are using `data_source.enabled` currently

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zengyan-amazon) [reply src/plugins/data_source/server/data_source_route_handler_context.ts:30]: index pattern service relies on the legacy client to fetch index mappings and fieldCaps, and we have most of our plugins using legacy client as well

Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_route_handler_context.ts:30]: I see. Created #2133 to track this.
By adding support for legacy client, I assume the route context for data source will look like this
```
datasource: {
  getClient：(id) => Promise<OpenSearchClient>
  legacy: {
    getClient: (id) => Promise<OpenSearchLegacyClient>
  }
}
```

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:79]: remove it

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:56]: let's follow the practice and put object creation into plugin constructor if possible

File: src/plugins/data_source/server/data_source_service.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/data_source_service.ts:56]: data source client and data source service relationship not clear to me. it seems client relies on the service, however, exposing this interface from the low level service class and letting service to pass itself (`this`) to client is kind of weird

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/data_source_route_handler_context.ts:51]: will this being executed for each request even though the request handler doesn't need to work with data source? please verify.

And let's try to keep the convention to use setup/start/stop to pass capabilities between components.

**Review by zengyan-amazon — CHANGES_REQUESTED** [review]

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts:56]: agreed, I'll move it

File: src/plugins/data_source/server/types.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/types.ts:13]: remove

File: src/plugins/data_source/server/data_source_service.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/data_source_service.ts:28]: you can pass config to setup function from plugin setup, instead of parsing config again here

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:63]: `dataSource` instead of `data_source`? please check the convention of other context fields

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:96]: let's make DataSourceClient a service style component, and exposes functions to plugin, instead of always creating new objects.

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (seraphjiang) [inline src/plugins/data_source/server/client/data_source_client.ts:36]: What does Ctor mean? could we use full name

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (seraphjiang) [inline src/plugins/data_source/server/client/data_source_client.ts:73]: add some log

File: src/plugins/data_source/server/data_source_service.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_service.ts]: Move logic from datasource client to datasource service, and expose functional interfaces only

File: src/plugins/data_source/server/data_source_route_handler_context.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_route_handler_context.ts]: addressed

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts:62]: yes it should be camel case

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/client/data_source_client.ts]: have done some refactor, log is added at the upper level,

File: src/plugins/data_source/server/client/data_source_client.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/client/data_source_client.ts]: removed in the new version

File: src/plugins/data_source/server/plugin.ts
Comment (zengyan-amazon) [inline src/plugins/data_source/server/plugin.ts:93]: this may cause 500 error, let's make a note here to handle that and returns 4xx as a follow up

**Review by zengyan-amazon — APPROVED** [review]
looks good

File: src/plugins/data_source/server/plugin.ts
Comment (zhongnansu) [reply src/plugins/data_source/server/plugin.ts:93]: Sure, I'll track that

**Review by seraphjiang — APPROVED** [review]
look forward for more coverage

**Comment by zhongnansu** [comment]
> look forward for more coverage

definitely, the existing tests can pass now. I am merging in to unblock others. Tracking UT improvement here for this module.
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2163

File: src/core/server/opensearch/client/configure_client.test.mocks.ts
Comment (zhongnansu) [inline src/core/server/opensearch/client/configure_client.test.mocks.ts:38]: by accident, will remove

**Pull request merged**
